torch/distributed/_shard/sharded_tensor/_ops/linear.py

```diff
@@ -328,19 +325,16 @@ def _handle_row_wise_sharding_sharded_tensor(
     """
     results = []
     local_shard = input.local_shards()[0].tensor
-    indices = [0] * world_size
-    reaggrance_partial = False
-    for idx, placement in enumerate(input._sharding_spec.placements):
-        indices[placement.rank()] = idx
-        if idx != placement.rank():
-            reaggrance_partial = True
+    if input.sharding_spec().dim not in (-1, len(input.size()) - 1):
+        raise NotImplementedError(
+            "The case when the input does not come from col-wise sharded "
+            "linear is not supported for row-wise sharded linear."
+        )
```

Comment:
Does this end up breaking any existing supported use cases? Also, why do we need to remove the rearrangement logic we had before?

Reply:
This does not break any existing use case.

Because partial tensor does not contain any sharding info. So we use the natural order from rank 0 to N. And no rearrange is needed. I tested both sharded linear and megatron prototype, they both passed.